### PR TITLE
Thrift to Metadata filter proto design

### PR DESCRIPTION
**Pull request by JuniorHsu**
Commit Message:
For [apache thrift](https://github.com/apache/thrift) compatible HTTP requests and responses, this filter parses the thrift metadata and put them into filter dynamic metadata for other filter usage. 

This is the initial proto design, which refers to other filters like json_to_metadata and payload_to_metadata.

Additional Description:
Risk Level: low
Testing: build
Docs Changes: yes
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue] #29371
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Comment by JuniorHsu** [comment]
@htuch could i have an early api review? the test failure is coverage which is expected

File: docs/root/configuration/http/http_filters/thrift_to_metadata_filter.rst [hunk 1]
```diff
@@ -0,0 +1,49 @@
+.. _config_http_filters_thrift_to_metadata:
+
+Envoy Thrift-To-Metadata Filter
```
Comment (htuch) [inline docs/root/configuration/http/http_filters/thrift_to_metadata_filter.rst:3]: Can you explain the relationship between this and the existing Thrift proxy support in Envoy? When does someone use one vs. the other? @zuercher

Comment (JuniorHsu) [reply docs/root/configuration/http/http_filters/thrift_to_metadata_filter.rst:3]: Sure. 

First of all, `thrift_proxy` is a network filter, and `thrift_to_metadata` is an HTTP filter.

`thrift_proxy` supports thrift RPC in HTTP 1.1 model, i.e., send a thrift serialized request over TCP and wait for the response. It's required to receive full response before send the next request. An example is [Apache THeader Transport](https://github.com/apache/thrift/blob/master/lib/py/src/transport/THeaderTransport.py)

`thrift_to_metadata` supports thrift serialized payload over HTTP, i.e., put RPC bytes in the POST body. Therefore, we could exploit all existing HTTP features and HTTP envoy filter. Moreover, we could utilize HTTP2 or HTTP3, which brings multiplexing and HOL avoidance. An example is [Apache THTTP](https://github.com/apache/thrift/blob/master/lib/py/src/transport/THttpClient.py)

While we use envoy HCM to proxy THTTP, it's natural to have a HTTP filter to extract thrift metadata, e.g., method name, as we do in `thrift_proxy`, which motivates us to implement this filter.

Moreover, we want to support the feature `payload_to_metadata` to extract thrift body information for other filters usage, but it won't be implemented at the first version so I mark it wip in the proto.

Comment (htuch) [reply docs/root/configuration/http/http_filters/thrift_to_metadata_filter.rst:3]: OK, makes sense. @zuercher can you do primary review on this?

**Comment by phlax** [comment]
@zuercher is it ok to assign this one to you?

File: api/envoy/extensions/filters/http/thrift_to_metadata/v3/thrift_to_metadata.proto [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+option java_outer_classname = "ThriftToMetadataProto";
+option java_multiple_files = true;
+option go_package = "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/http/thrift_to_metadata/v3;thrift_to_metadatav3";
+option (udpa.annotations.file_status).package_version_status = ACTIVE;
+option (xds.annotations.v3.file_status).work_in_progress = true;
+
+// [#protodoc-title: Thrift-To-Metadata Filter]
+//
+// The configuration for transforming thrift metadata into filter metadata. This is useful
+// for matching load balancer subsets, logging, etc.
```
Comment (zuercher) [inline api/envoy/extensions/filters/http/thrift_to_metadata/v3/thrift_to_metadata.proto:24]: Can we clarify that this is specifically for Thrift-over-HTTP?

File: api/envoy/extensions/filters/http/thrift_to_metadata/v3/thrift_to_metadata.proto [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+  // The Thrift reply type, string value. This is only valid for response rules. Values are "success" and "error".
+  REPLY_TYPE = 6;
+}
+
+message KeyValuePair {
+  // The namespace — if this is empty, the filter's namespace will be used.
+  string metadata_namespace = 1;
+
+  // The key to use within the namespace.
+  string key = 2 [(validate.rules).string = {min_len: 1}];
```
Comment (zuercher) [inline api/envoy/extensions/filters/http/thrift_to_metadata/v3/thrift_to_metadata.proto:57]: IIRC the metadata allows nesting of keys. Is that supported here? If yes, please document how (I assume period-delimited keys). It not, should it be?

File: docs/root/configuration/http/http_filters/thrift_to_metadata_filter.rst [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+===============================
+* This filter should be configured with the type URL ``type.googleapis.com/envoy.extensions.filters.http.thrift_to_metadata.v3.ThriftToMetadata``.
+* :ref:`v3 API reference <envoy_v3_api_msg_extensions.filters.http.thrift_to_metadata.v3.ThriftToMetadata>`
+
+This filter is configured with rules that will be matched against Apache thrift compatible requests and responses.
+The filter will parse the thrift body, extract thrift metadata or payload, and add them to dynamic metadata
+based on the configuration of the rule.
```
Comment (zuercher) [inline docs/root/configuration/http/http_filters/thrift_to_metadata_filter.rst:10]: Again, make it clear that this if for thrift requests/responses within the context of HTTP.

**Review by zuercher — COMMENTED** [review]
As a design this makes sense to me. Given that there's no implementation here, do you want to mark all the protobuf as not yet implemented and move forward with this PR just to get the protobuf API in place?

File: api/envoy/extensions/filters/http/thrift_to_metadata/v3/thrift_to_metadata.proto (hunk 3 above)
Comment (JuniorHsu) [reply api/envoy/extensions/filters/http/thrift_to_metadata/v3/thrift_to_metadata.proto:57]: hmm i have little knowledge on nesting of keys here but I was following prior art, e.g., http/header_to_metadata filter.

https://github.com/envoyproxy/envoy/blob/9536ce2d3fdb234f8006ba4092e59577c15ea179/api/envoy/config/core/v3/base.proto#L248

Here's is the namespace to struct mapping, and we have `key` `value` mapping inside of the struct.

**Comment by JuniorHsu** [comment]
> do you want to mark all the protobuf as not yet implemented and move forward with this PR just to get the protobuf API in place

Sure I'll have a very basic unit test to pass the coverage so we can move on the wip filter. Thanks for the great idea.

**Comment by JuniorHsu** [comment]
> do you want to mark all the protobuf as not yet implemented

just notice i already set `option (xds.annotations.v3.file_status).work_in_progress = true;`

**Comment by JuniorHsu** [comment]
@zuercher I addressed all the comments except this https://github.com/envoyproxy/envoy/pull/33919#discussion_r1599303324 which needs some input. Also add some trivial unit test, hopefully pass the coverage test

File: api/envoy/extensions/filters/http/thrift_to_metadata/v3/thrift_to_metadata.proto (hunk 3 above)
Comment (zuercher) [reply api/envoy/extensions/filters/http/thrift_to_metadata/v3/thrift_to_metadata.proto:59]: Ok. I suppose in this case it's probably better to be consistent.

**Review by zuercher — APPROVED** [review]
This looks good to me. I think it needs approval from someone on @envoyproxy/api-shepherds.

**Comment by zuercher** [comment]
> > do you want to mark all the protobuf as not yet implemented
> 
> just notice i already set `option (xds.annotations.v3.file_status).work_in_progress = true;`

Sorry -- I didn't realize that could be done at the file level vs. per-message.

File: api/envoy/extensions/filters/http/thrift_to_metadata/v3/thrift_to_metadata.proto [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+message Rule {
+  // The field to match on. If set, takes precedence over field_selector.
+  Field field = 1;
+
+  // Specifies that a match will be performed on the value of a field. If set,
+  // the whole http body will be buffered to extract the field value, which
+  // may have performance implications.
+  FieldSelector field_selector = 2 [(xds.annotations.v3.field_status).work_in_progress = true];
+
+  // The key-value pair to set in the metadata if the field is present.
```
Comment (htuch) [inline api/envoy/extensions/filters/http/thrift_to_metadata/v3/thrift_to_metadata.proto:90]: Can you disambiguate everywhere in this proto between Thrift and Envoy metadata? It's a bit confusing to the casual reader. So just write "thrift metadata" here for example.

File: api/envoy/extensions/filters/http/thrift_to_metadata/v3/thrift_to_metadata.proto [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+//
+// The Thrift to Metadata filter serves for thrift over HTTP traffic, expecting serialized
+// Thrift request and response bodies in the HTTP payload. It extracts thrift metadata from the
+// HTTP body and put them into the filter metadata. This is useful for matching load balancer
+// subsets, logging, etc.
+//
+// Thrift to Metadata :ref:`configuration overview <config_http_filters_thrift_to_metadata>`.
+// [#extension: envoy.filters.http.thrift_to_metadata]
+
+enum Field {
```
Comment (htuch) [inline api/envoy/extensions/filters/http/thrift_to_metadata/v3/thrift_to_metadata.proto:31]: Should this live in core `netowrk.thrift.v3` with other defs? I.e. are these enum values defined by the Thrift standard or us? @zuercher

File: api/envoy/extensions/filters/http/thrift_to_metadata/v3/thrift_to_metadata.proto [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+  string key = 2 [(validate.rules).string = {min_len: 1}];
+
+  // When used for on_present case, if value is non-empty it'll be used instead
+  // of the field value.
+  //
+  // When used for on_missing case, a non-empty value must be provided.
+  google.protobuf.Value value = 3;
+}
+
+message FieldSelector {
```
Comment (htuch) [inline api/envoy/extensions/filters/http/thrift_to_metadata/v3/thrift_to_metadata.proto:68]: Can you add references to the relevant part of the Thrift spec?

File: api/envoy/extensions/filters/http/thrift_to_metadata/v3/thrift_to_metadata.proto [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+  // The field to match on. If set, takes precedence over field_selector.
+  Field field = 1;
+
+  // Specifies that a match will be performed on the value of a field. If set,
+  // the whole http body will be buffered to extract the field value, which
+  // may have performance implications.
+  FieldSelector field_selector = 2 [(xds.annotations.v3.field_status).work_in_progress = true];
+
+  // The key-value pair to set in the metadata if the field is present.
+  KeyValuePair on_present = 3;
```
Comment (htuch) [inline api/envoy/extensions/filters/http/thrift_to_metadata/v3/thrift_to_metadata.proto:91]: Confusingly, when looking at header_to_metadatea.proto, this actually uses the `value` to override, despite the documentation of the `value` field. Did you notice that discrepancy? What is the plan in this filter?

Comment (JuniorHsu) [reply api/envoy/extensions/filters/http/thrift_to_metadata/v3/thrift_to_metadata.proto:91]: We plan to do the same with `header_to_metadata` so I'll probably update the comment.

If the `value` is present in `on_present` and field is also present (e.g., we have method_name in thrift metadata), we put on_present.value.

If the `value` is not present in `on_present`, we put method_name from thrift metadata for example

File: api/envoy/extensions/filters/http/thrift_to_metadata/v3/thrift_to_metadata.proto (hunk 6 above)
Comment (JuniorHsu) [reply api/envoy/extensions/filters/http/thrift_to_metadata/v3/thrift_to_metadata.proto:31]: i can answer the enum part. 

these values are referred from https://github.com/envoyproxy/envoy/blob/main/source%2Fextensions%2Ffilters%2Fnetwork%2Fthrift_proxy%2Fmetadata.h#L51-L57 which is inferred or extract from the thrift message of different transport/protocol we support.

spec doesn't have definition of `metadata`. `metadata` is kinda our derivative or thrift and we gives those information for thrift filters.

For example, this is the place we extract seq_id and method name from binary protocol
https://github.com/apache/thrift/blob/master/doc/specs/thrift-binary-protocol.md#message

header flags lives only in header transport, so will be absl::nullopt for other transport
https://github.com/apache/thrift/blob/master/doc/specs/HeaderFormat.md

thrift message itself doesn't specify it's transport, so we either "know" this before we extract, or we determine by magic prefix bytes.

File: api/envoy/extensions/filters/http/thrift_to_metadata/v3/thrift_to_metadata.proto (hunk 7 above)
Comment (JuniorHsu) [reply api/envoy/extensions/filters/http/thrift_to_metadata/v3/thrift_to_metadata.proto:72]: [apache spec](https://thrift.apache.org/docs/idl) is hard to read at least for me :( I link the prior-art and also add one more example hopefully it'll be more clear.

File: api/envoy/extensions/filters/http/thrift_to_metadata/v3/thrift_to_metadata.proto (hunk 5 above)
Comment (JuniorHsu) [reply api/envoy/extensions/filters/http/thrift_to_metadata/v3/thrift_to_metadata.proto]: done. thanks

File: docs/root/configuration/http/http_filters/thrift_to_metadata_filter.rst (hunk 1 above)
Comment (JuniorHsu) [reply docs/root/configuration/http/http_filters/thrift_to_metadata_filter.rst:3]: done here
https://github.com/envoyproxy/envoy/pull/33919#pullrequestreview-2072613349

**Comment by JuniorHsu** [comment]
we got approval and api reviewed so i think we're good to merge.
https://github.com/envoyproxy/envoy/pull/33919#pullrequestreview-2072613349

**Review by htuch — APPROVED** [review]

**Pull request merged**